### PR TITLE
feat(test): Add project counter test helpers

### PR DESCRIPTION
**Pull request by lobsterkatie**
This adds two test helper functions, `patch_group_creation` and `create_existing_group`, to use in project counter tests. 

The former patches `Group.objects.create` in order to both spy on it and collect its results. The latter does that patching and then saves an event with a message designed to be different from existing groups' messages, so as to guarantee we'll create a new group. In that process, `Group.objects.create` runs, allowing us to make assertions about what it does.`create_existing_group` also includes some assertions to make sure that things run according to plan. 

Finally, this PR adds a simple happy path test using the helpers.

### CODE REVIEW

File: tests/sentry/models/test_projectcounter.py [hunk 1]
```diff
@@ -14,3 +20,43 @@ def test_increment(default_project, upsert_sample_rate):
...
+
+    yield (group_creation_spy, group_creation_results)
+
+
+def create_existing_group(project, monkeypatch, message):
+    with patch_group_creation(monkeypatch) as patches:
+        group_creation_spy, group_creation_results = patches
+
+        event = save_new_event({"message": message}, project)
+        group = Group.objects.filter(id=event.group_id).first()
```
Comment (wedamija) [inline tests/sentry/models/test_projectcounter.py:41]: Nit: Should we just use `.get()` here?

File: tests/sentry/models/test_projectcounter.py [hunk 2]
```diff
@@ -14,3 +20,43 @@ def test_increment(default_project, upsert_sample_rate):
...
+
+        assert (
+            group.times_seen == 1
+        ), "Error: No new group was created. This is probably because the given message matched that of an existing group."
+
+        assert group_creation_spy.call_count == 1
+        assert group_creation_results[0] == group
+
+        # This equality ensures the next new group's `short_id` won't conflict with group's `short_id`.
+        counter = Counter.objects.filter(project_id=project.id).first()
```
Comment (wedamija) [inline tests/sentry/models/test_projectcounter.py:51]: Nit: Should we use`.get()` here too?

**Review by wedamija — APPROVED** [review]

File: tests/sentry/models/test_projectcounter.py (hunk 1 above)
Comment (lobsterkatie) [reply tests/sentry/models/test_projectcounter.py:41]: I thought that migration we worked on failed because we were using `.get()`. Am I remembering wrong?

**Comment by vartec** [comment]
This returns a static integer, so there's a change of a race condition, i.e. when you use it to update Counter another group was create concurrently and it's invalid number again. 

IMO the approach should be this returning a [SubQuery expression](https://docs.djangoproject.com/en/5.0/ref/models/expressions/#subquery-expressions) and an [F() expression](https://docs.djangoproject.com/en/5.0/ref/models/expressions/#f-expressions) to add 1.

In fact I think that using [OuterRef](https://docs.djangoproject.com/en/5.0/ref/models/expressions/#django.db.models.OuterRef) this could be a completely static expression. 

Basically that the resulting SQL would be something like 
```sql
UPDATE sentry_projectcounter 
SET value = (
   SELECT max(short_id) + 1 
   FROM sentry_groupedmessage
   WHERE sentry_groupedmessage.project_id = sentry_projectcounter.project_id
)
```

File: tests/sentry/models/test_projectcounter.py (hunk 1 above)
Comment (wedamija) [reply tests/sentry/models/test_projectcounter.py:41]: You're remembering right, but the context around that was that we had hardcoded IDs that didn't exist locally, so the exception being thrown caused the failure.

If the group doesn't exist here, we'll also have a bug because we're not checking for None. .get() is probably most correct here

Comment (lobsterkatie) [reply tests/sentry/models/test_projectcounter.py:41]: Ah, I see. All right, `.get()` it is!

File: tests/sentry/models/test_projectcounter.py (hunk 2 above)
Comment (lobsterkatie) [reply tests/sentry/models/test_projectcounter.py:51]: I switched this one also.

**Review by vartec — APPROVED** [review]
